drivers/source/I2C.cpp

```diff
@@ -72,26 +65,13 @@ void I2C::frequency(int hz)
     // We want to update the frequency even if we are already the bus owners
     i2c_frequency(&_i2c, _hz);
 
-    // Updating the frequency of the bus we become the owners of it
-    _owner = this;
-    unlock();
-}
-
-void I2C::aquire()
-{
-    lock();
-    if (_owner != this) {
-        i2c_frequency(&_i2c, _hz);
-        _owner = this;
-    }
     unlock();
 }
 
 // write - Master Transmitter Mode
 int I2C::write(int address, const char *data, int length, bool repeated)
 {
     lock();
-    aquire();
```

Comment:
removing `aquire` from write/read and transfer methods might break applications or it does not? If owner != this, freq() would be invoked and we are removing this case completely. Is it intention and actually fixes the issue ?

Reply:
This change would potentially break applications that are using multiple `I2C` objects to represent the same bus. Whether they should ever have been doing that is open to debate.

Reply:
> Whether they should ever have been doing that is open to debate.

Someone from @ARMmbed/mbed-os-hal could help reviewing this and find a solution for now and the next major version

If this is a breaking change (not just functionality change), we might want to hide this behind a configuration and do this breaking change for Mbed 7. all changes are related so it's not just acquire. Or any other options?

Config option could be `i2c-aquire-enabled`, set to true by default.

Reply:
Config option is probably a good idea.

Not sure about naming. I don't think you can use the misspelt name of a non-public method though :)

A few more suggestions, for either polarity - `i2c-fully-mutually-exclusive` / `i2c-independent-objects` / `i2c-require-one-object-per-peripheral` / `i2c-permit-multiple-objects-per-peripheral'?

Whatever you choose, it's going to need a slightly long description as well to spell it out. I can't see a way of making the name self-explanatory.

Reply:
Is there a planned release date for Mbed 7?

If i get it right using this fix we just need to reconfigure the I2C object before using a sensor that needs another frequency/I2C config than the sensor we communicated with last time on the same port. If that would be all that is there to it, i would not hesitate to merge this fix into the next minor release.

A config option is a good compromise, however that way Mbed would default to an I2C implementation that puts a huge load on the MCU that can result in several unspecific symptoms which are not straightforward to find the root cause of (just finding the root would probably take more time than reading the porting guide of the next release, understanding the flaw in I2C that has been addressed and change code). 

As for config options names: the proposed names are hard to remember and i do not see people wanting to use more than 1 I2C implementation. So why not just call it `i2c-high-processor-impact` and document clearly that an identified and fixed bug causes high mcu load but still leave space for the developer to decide if they want to struggle with a buggy implementation when they finally realize it eats up much of their MCU time...or they prefer using a fixed implementation that allows them to use the MCU for useful stuff 😎

Reply:
Hi @0xc0170,

>  might break applications or it does not?

This is a breaking change only in case when two or more I2C instances simultaneously exist for a single physical bus. I'm not sure that such design is valid, see the [issue](https://github.com/ARMmbed/mbed-os/issues/14004)

As @kjbracey-arm said, 

> Whether they should ever have been doing that is open to debate

I fully agree. I also think that config option is a good idea. I would say `i2c-one-instance-per-bus`.

Reply:
> an I2C implementation that puts a huge load on the MCU

I'd say that issue remains and is worth investigating - is the STM HAL taking longer to reconfigure that it should? If someone is really needing to reconfigure a bus, this could still be an issue. However...

> another frequency/I2C config than the sensor we communicated with last time on the same port.

Is this a real-life use case? Devices that need to be spoken to at a lower frequency, but can cope with you speaking to other devices at a higher speed without getting confused? Don't you have to run the whole bus at the signalling rate of the slowest device on it?

Reply:
> I'd say that issue remains and is worth investigating - is the STM HAL taking longer to reconfigure that it should?

I have no idea as i do not have any STM targets.

> Is this a real-life use case?
Yes, theoretically it is possible if you are lucky enough and your slow device does not begin to output erronous data when it gets confused. Though you are right that it should be avoided in practice as you can never tell in advance if your slow device would begin to misbehave or not.

Reply:
If all agree, this would be feature type update - functionality change. Fixing the broken premise we did with I2c - no breaking change. No config, we will need to document this in the release notes.

Impact of changes and Migration guide would need more details to understand the scope of this change.

Let's wait for @ARMmbed/mbed-os-hal to review